stl/inc/xmemory

```diff
@@ -1033,147 +1034,190 @@ struct _Iterator_base0 {
 
-    const _Container_base12* _Mycont;
-    _Iterator_base12* _Myfirstiter;
+    const _Container_base12* _Mycont       = nullptr;
+    mutable _Iterator_base12* _Myfirstiter = nullptr;
 };
 
 struct _Container_base12 {
 public:
-    _Container_base12() noexcept : _Myproxy(nullptr) {}
+    _CONSTEXPR20_DYNALLOC _Container_base12() noexcept = default;
 
     _Container_base12(const _Container_base12&) = delete;
     _Container_base12& operator=(const _Container_base12&) = delete;
 
-    void _Orphan_all() noexcept;
-    void _Swap_proxy_and_iterators(_Container_base12&) noexcept;
+    _CONSTEXPR20_DYNALLOC void _Orphan_all() noexcept;
+    _CONSTEXPR20_DYNALLOC void _Swap_proxy_and_iterators(_Container_base12&) noexcept;
 
     template <class _Alloc>
-    void _Alloc_proxy(_Alloc&& _Al) {
+    _CONSTEXPR20_DYNALLOC void _Alloc_proxy(_Alloc&& _Al) {
         _Container_proxy* const _New_proxy = _Unfancy(_Al.allocate(1));
         _Construct_in_place(*_New_proxy, this);
         _Myproxy            = _New_proxy;
         _New_proxy->_Mycont = this;
     }
 
     template <class _Alloc>
-    void _Reload_proxy(_Alloc&& _Old_alloc, _Alloc&& _New_alloc) {
+    _CONSTEXPR20_DYNALLOC void _Reload_proxy(_Alloc&& _Old_alloc, _Alloc&& _New_alloc) {
         // pre: no iterators refer to the existing proxy
         _Container_proxy* const _New_proxy = _Unfancy(_New_alloc.allocate(1));
         _Construct_in_place(*_New_proxy, this);
         _New_proxy->_Mycont = this;
         _Delete_plain_internal(_Old_alloc, _STD exchange(_Myproxy, _New_proxy));
     }
 
-    _Container_proxy* _Myproxy;
+    _Container_proxy* _Myproxy = nullptr;
+
+private:
+    _CONSTEXPR20_DYNALLOC void _Orphan_all_unlocked() noexcept;
+    inline void _Orphan_all_locked() noexcept;
+    _CONSTEXPR20_DYNALLOC void _Swap_proxy_and_iterators_unlocked(_Container_base12&) noexcept;
+    inline void _Swap_proxy_and_iterators_locked(_Container_base12&) noexcept;
 };
 
 struct _Iterator_base12 { // store links to container proxy, next iterator
-    _Iterator_base12() noexcept : _Myproxy(nullptr), _Mynextiter(nullptr) {} // construct orphaned iterator
+public:
+    _CONSTEXPR20_DYNALLOC _Iterator_base12() noexcept = default; // construct orphaned iterator
 
-    _Iterator_base12(const _Iterator_base12& _Right) noexcept : _Myproxy(nullptr), _Mynextiter(nullptr) {
+    _CONSTEXPR20_DYNALLOC _Iterator_base12(const _Iterator_base12& _Right) noexcept {
         *this = _Right;
     }
 
-    _Iterator_base12& operator=(const _Iterator_base12& _Right) noexcept {
+    _CONSTEXPR20_DYNALLOC _Iterator_base12& operator=(const _Iterator_base12& _Right) noexcept {
         if (_Myproxy != _Right._Myproxy) {
             if (_Right._Myproxy) {
                 _Adopt(_Right._Myproxy->_Mycont);
             } else { // becoming invalid, disown current parent
 #if _ITERATOR_DEBUG_LEVEL == 2
-                _Lockit _Lock(_LOCK_DEBUG);
                 _Orphan_me();

```

Comment:
There appears to be an ABI/mix-and-match issue here. The problem is that `_Orphan_me`'s semantics are being changed - previously, it expected the debug lock to be held, and now it is responsible for acquiring the debug lock. If `_Iterator_base12::operator=`'s new definition (which doesn't acquire the lock) is mixed with `_Orphan_me`'s old definition (which didn't acquire the lock), there will be doom. I believe we can avoid this by renaming `_Orphan_me`, so that if the linker looks for the old name, it will find the old definition, and similarly for the new name and new definition. `operator=` will have a varying definition, but that's okay, because whichever version is selected, it will provide the same observable behavior.

No strong opinions on the new name; could use a `_v2` suffix, or something else.

Reply:
Added a _v2 suffix

Reply:
Please shout if you meant to duplicate the existing functionality

Reply:
The rename looks good - this is equivalent to deleting the old function and adding the new functions all with different names, which is ABI-safe. (No functions are keeping their names while changing semantics in an incompatible way.) Thanks!